### PR TITLE
Remove preprinter, add parenthesizer callback to emit

### PR DESCRIPTION
**Pull request by rbuckton**
This reverts the "preprinter" transform from #42676 that was primarily used to ensure correct parenthesization using the node factory and amends the emitter to support providing a parenthesizer rule callback when emitting certain nodes.

Related #43486

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot perf test

**Comment by rbuckton** [comment]
@typescript-bot perf test

**Comment by andrewbranch** [comment]
@rbuckton can you point out what code is new, as opposed to a reversion?

**Comment by rbuckton** [comment]
In _emitter.ts_, most of the changes are a reversion. I left in some of the changes in comment/sourcemap emit as it improves the number of cases where we can use our binary expression trampoline to avoid exhausting the call stack.

The only real *addition* is the ability to pass a `parenthesizerRule` to `emit`/`emitExpression`, which accomplishes the same thing I was trying to fix when I added the `preprinter` transform, but without the extra full-walk of the tree.

Anything in _nodeFactory.ts_, _parenthesizerRules.ts_, and _types.ts_ is new and was added to support passing a `parenthesizerRule` in _emitter.ts_.

**Comment by rbuckton** [comment]
This one wasn't as simple as just rolling back the change, as the work to create a reusable binary expression trampoline is something we want to keep so that we're consistent across the binder, checker, and emitter. As a result, we also kept the changes to comment and source map emit so that we could reuse the logic inside the binary expression trampoline.

File: src/compiler/types.ts [hunk 1]
```diff
@@ -6785,6 +6785,8 @@ namespace ts {
 
     /* @internal */
     export interface ParenthesizerRules {
+        getParenthesizeLeftSideOfBinaryForOperator(binaryOperator: SyntaxKind): (leftSide: Expression) => Expression;
+        getParenthesizeRightSideOfBinaryForOperator(binaryOperator: SyntaxKind): (rightSide: Expression) => Expression;
```
Comment (rbuckton) [inline src/compiler/types.ts:6789]: These are new, and exist to support passing in a `(node: Node) => Node` parenthesizer rule for binary expressions. The implementation simply wraps `parenthesizeLeftSideOfBinary` or `parenthesizeRightSideOfBinary` and caches the result for reuse.

File: src/compiler/transformer.ts [hunk 2]
```diff
@@ -562,31 +562,31 @@ namespace ts {
     }
 
     export const nullTransformationContext: TransformationContext = {
-        get factory() { return factory; },
```
Comment (rbuckton) [inline src/compiler/transformer.ts:565]: This is simply a reordering of existing properties to match the order in `transformNodes` so they (hopefully) share the same "map" in V8.

File: tests/baselines/reference/inlineJsxFactoryDeclarationsLocalTypes.js [hunk 3]
```diff
@@ -161,4 +161,6 @@ var _brokenTree = (0, renderer_1.dom)(component_1.MySFC, { x: 1, y: 2 },
     (0, renderer_1.dom)(component_1.MyClass, { x: 3, y: 4 }),
     (0, renderer_1.dom)(component_1.MyClass, { x: 5, y: 6 }));
 // Should fail, nondom isn't allowed as children of dom
-var _brokenTree2 = (0, renderer_1.dom)(DOMSFC, { x: 1, y: 2 }, component_1.tree, component_1.tree);
+var _brokenTree2 = (0, renderer_1.dom)(DOMSFC, { x: 1, y: 2 },
```
Comment (rbuckton) [inline tests/baselines/reference/inlineJsxFactoryDeclarationsLocalTypes.js:164]: This indentation matches the emit in `release-4.2`.

File: src/compiler/emitter.ts [hunk 4]
```diff
@@ -1199,5559 +1224,4489 @@ namespace ts {
...
+            switch (phase) {
+                case PipelinePhase.Notification:
+                    if (onEmitNode !== noEmitNotification && (!isEmitNotificationEnabled || isEmitNotificationEnabled(node))) {
+                        return pipelineEmitWithNotification;
+                    }
+                    // falls through
+                case PipelinePhase.Substitution:
+                    if (substituteNode !== noEmitSubstitution && (lastSubstitution = substituteNode(emitHint, node) || node) !== node) {
+                        if (currentParenthesizerRule) {
+                            lastSubstitution = currentParenthesizerRule(lastSubstitution);
```
Comment (rbuckton) [inline src/compiler/emitter.ts:1343]: This (and the code path that sets `currentParenthesizerRule`) is the only "actual" change from what we previously were doing in 4.2. The point here is to only invoke the parenthesizer rule when substitution results in a *different* node (which might have a different precedence).

**Comment by rbuckton** [comment]
While not perfect (as there are some other changes in _emitter.ts_ unrelated to this), you can get a gist of what changed and was reverted by doing the following on the commandline:

```sh
git difftool d3b8d5e9c423d2565e9c5122326fbf4a2a2e5d6a^..substitutionParenthesizer -- src/compiler/emitter.ts
```

**Comment by rbuckton** [comment]
I'll actually push up a change that reorders some of the functions so that they line up with where they were prior to $43676 to make the command-line diff easier to read through.

**Comment by rbuckton** [comment]
I've pushed up a change that reorders some of the functions so that they line up with where they were prior to #43676 to make the command-line diff easier to read through.

**Review by andrewbranch — APPROVED** [review]
I’m not very familiar with this area of code, nor the changes that are being reverted, but the additions you pointed out make sense to me.

**Comment by rbuckton** [comment]
I don't think this should have broken anything, but just to be on the safe side...

@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Comment by rbuckton** [comment]
RWC tests failed, but its unclear why. I'm running them locally.
User tests were fine except for `fp-ts`, which for some reason ran out of heap memory. I will have to investigate that one to see what's causing the crash.

**Comment by rbuckton** [comment]
RWC failures are mostly unrelated to this change, the few that are are changes in indentation that were likely present *before* the preprinter transform was added (similar to what I mentioned [here](https://github.com/microsoft/TypeScript/pull/43652#discussion_r614486773)), so this is just going back to existing behavior.

**Comment by rbuckton** [comment]
The `fp-ts` user tests didn't crash locally. @weswigham did you mention you were you looking into an OOM?

**Comment by weswigham** [comment]
@sandersn did.a bit ago, iirc?

**Comment by rbuckton** [comment]
I cannot repro the OOM locally.

@typescript-bot user test this

**Comment by rbuckton** [comment]
Ok. It didn't crash that time. Must have been a glitch in the matrix.

**Comment by rbuckton** [comment]
Summary of DT failures:
- There's a crash that doesn't list the project that is crashing, however this is also occurring in nightly, so I don't think its related to this change: https://typescript.visualstudio.com/TypeScript/_build/results?buildId=101141&view=logs&j=ae5f67bc-b508-52d0-18c7-3e0886a35c88&t=344fa42e-5be7-595c-913e-e7acf0f0490f&l=13595
- `mithril` is crashing due to running out of stack space in checker, which doesn't seem to be related to this change as far as I can tell. It occurs in the nightly as well: https://typescript.visualstudio.com/TypeScript/_build/results?buildId=101141&view=logs&j=abaf456e-5a9c-50a3-94a8-5c090880e34b&t=94425fb2-771d-59ba-1fed-934966b6530f&l=13594
- `electron-notifications`, `electron-notify`, and `electron-json-storage` are having issues with `traceDeprecation` in `../node/process.d.ts`, but this occurs in the nightly as well, so is unrelated to this change: 
    - `electron-notifications` - https://typescript.visualstudio.com/TypeScript/_build/results?buildId=101141&view=logs&j=71179031-6417-5a2f-3d87-af6fce2011e4&t=5ac81f15-0a74-5a9b-b155-06ef07cd84f6&l=13596
    - `electron-json-storage` - https://typescript.visualstudio.com/TypeScript/_build/results?buildId=101141&view=logs&j=ae9e61d0-c116-562e-ef72-0c532f74ef92&t=ccd97d0d-fa78-59da-ae5f-9c54639b4c00&l=13600
    - `electron-notify` - https://typescript.visualstudio.com/TypeScript/_build/results?buildId=101141&view=logs&j=ae9e61d0-c116-562e-ef72-0c532f74ef92&t=ccd97d0d-fa78-59da-ae5f-9c54639b4c00&l=13634
- `cytoscape` is crashing due to running out of stack space in checker, which is also occurring in the nightly and is unrelated: https://typescript.visualstudio.com/TypeScript/_build/results?buildId=101141&view=logs&j=ae9e61d0-c116-562e-ef72-0c532f74ef92&t=ccd97d0d-fa78-59da-ae5f-9c54639b4c00&l=13588

None of the DT failures seem to be related to this change, but warrant further investigation outside of this PR.

**Comment by sandersn** [comment]
The other crash has the same origin; the electron-* are known failures that shouldn't show up in the final output. They should, I hope, be gone after the additional `grep` that @weswigham and I stuck onto the end of the `diff`.

**Pull request merged**